server/server/src/main/java/org/infinispan/server/server/ExitHandler.java

```diff
@@ -0,0 +1,26 @@
+package org.infinispan.server.server;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+
+/**
+ * ExitHandler provides alternate implementations for {@link System#exit(int)} to use in different scenarios (e.g.
+ * tests)
+ *
+ * @author Tristan Tarrant
+ * @since 10.0
+ */
+
+public abstract class ExitHandler {
+   CompletableFuture<Integer> exitFuture;
+
+   ExitHandler() {
+      exitFuture = new CompletableFuture<>();
+   }
+
+   public void waitForExit() throws ExecutionException, InterruptedException {
```

Comment:
I am not a fan of the blocking for this. Surely we can expose a callback like `BiConsumer<Integer, Throwable>` and register it with the `exitFuture`?

Reply:
We need a way for the server to do a "wait until exit". How would you implement that ?

Reply:
I was saying that it could be nonblocking is all. Similar issue to https://github.com/infinispan/infinispan/pull/6497#discussion_r241077563.

This isn't yet being invoked from what I can see. Given that we could probably revisit this later if needed, and it probably would clear up some questions.